drivers/input/input_paw32xx.c

```diff
@@ -0,0 +1,443 @@
+	int ret;
+
+	ret = paw32xx_read_reg(dev, PAW32XX_PRODUCT_ID1, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (val != PRODUCT_ID_PAW32XX) {
+		LOG_ERR("Invalid product id: %02x", val);
+		return -ENOTSUP;
+	}
+
+	ret = paw32xx_update_reg(dev, PAW32XX_CONFIGURATION,
+				 CONFIGURATION_RESET, CONFIGURATION_RESET);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_sleep(K_MSEC(RESET_DELAY_MS));
+
+	if (cfg->invert_x || cfg->invert_y) {
+		ret = paw32xx_write_reg(dev, PAW32XX_WRITE_PROTECT, WRITE_PROTECT_DISABLE);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = paw32xx_read_reg(dev, PAW32XX_MOUSE_OPTION, &val);
+		if (ret < 0) {
+			return ret;
+		}
+
+		WRITE_BIT(val, MOUSE_OPTION_MOVX_INV_BIT, cfg->invert_x);
+		WRITE_BIT(val, MOUSE_OPTION_MOVY_INV_BIT, cfg->invert_y);
+
+		ret = paw32xx_write_reg(dev, PAW32XX_MOUSE_OPTION, val);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = paw32xx_write_reg(dev, PAW32XX_WRITE_PROTECT, WRITE_PROTECT_ENABLE);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	if (cfg->res_cpi > 0) {
+		paw32xx_set_resolution(dev, cfg->res_cpi);
+	}
+
+	paw32xx_force_awake(dev, cfg->force_awake);
+
+	return 0;
+}
+
+static int paw32xx_init(const struct device *dev)
+{
+	const struct paw32xx_config *cfg = dev->config;
+	struct paw32xx_data *data = dev->data;
+	int ret;
+
+	if (!spi_is_ready_dt(&cfg->spi)) {
+		LOG_ERR("%s is not ready", cfg->spi.bus->name);
+		return -ENODEV;
+	}
+
+	data->dev = dev;
+
+	k_work_init(&data->motion_work, paw32xx_motion_work_handler);
+
+	if (!gpio_is_ready_dt(&cfg->motion_gpio)) {
+		LOG_ERR("%s is not ready", cfg->motion_gpio.port->name);
+		return -ENODEV;
+	}
+
+	ret = gpio_pin_configure_dt(&cfg->motion_gpio, GPIO_INPUT);
+	if (ret != 0) {
+		LOG_ERR("Motion pin configuration failed: %d", ret);
+		return ret;
+	}
+
+	gpio_init_callback(&data->motion_cb, paw32xx_motion_handler,
+			   BIT(cfg->motion_gpio.pin));
+
+	ret = gpio_add_callback_dt(&cfg->motion_gpio, &data->motion_cb);
+	if (ret < 0) {
+		LOG_ERR("Could not set motion callback: %d", ret);
+		return ret;
+	}
+
+	ret = paw32xx_configure(dev);
+	if (ret != 0) {
+		LOG_ERR("Device configuration failed: %d", ret);
+		return ret;
+	}
+
+	ret = gpio_pin_interrupt_configure_dt(&cfg->motion_gpio,
+					      GPIO_INT_EDGE_TO_ACTIVE);
+	if (ret != 0) {
+		LOG_ERR("Motion interrupt configuration failed: %d", ret);
+		return ret;
+	}
+
+	ret = pm_device_runtime_enable(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to enable runtime power management: %d", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int paw32xx_pm_action(const struct device *dev,
+			     enum pm_device_action action)
+{
+	int ret;
+	uint8_t val;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		val = CONFIGURATION_PD_ENH;
+		break;
+	case PM_DEVICE_ACTION_RESUME:
+		val = 0;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	ret = paw32xx_update_reg(dev, PAW32XX_CONFIGURATION,
+				 CONFIGURATION_PD_ENH, val);
+	if (ret < 0) {
+		return ret;
```

Comment:
Nit: just return ret here, if statement not needed

Reply:
not a fan of that pattern to be totally honest, I always find myself hacking something in the code and having to add the explicit return so I can, for example, print a value after the function